src/dataflow/src/source/postgres.rs

```diff
@@ -157,12 +169,34 @@ impl PostgresSourceReader {
                         let a: Datum = row.get(n).into();
                         a
                     }));
-                    snapshot_tx.insert(mz_row).await?;
+                    try_recoverable!(snapshot_tx.insert(mz_row.clone()).await);
+                    try_fatal!(serde_json::to_writer(&mut buffer, &mz_row));
+                }
+            }
+            try_fatal!(buffer.flush());
```

Comment:
The producer code can error out before this line which will cause the data to be lost for the underlying writer. I think what we need is to create a `BufReader<BufWriter<File>>` in `start()` (once this https://github.com/tokio-rs/tokio/pull/3810 is released we can switch to a `BufStream<File>`) and make `produce_snapshot` and `revert_snapshot` take a generic `W: AsyncWriter` and `R: AsyncReader` respectively.

In this manner we never have to care about calling `flush` as the temporary file will remain wrapper for the whole time